flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/UnreachableCoalesceArgumentsRemoveRule.java

```diff
@@ -0,0 +1,197 @@
+            Config.EMPTY.as(Config.class).withProject().toRule();
+    public static final RelOptRule FILTER_INSTANCE =
+            Config.EMPTY.as(Config.class).withFilter().toRule();
+    public static final RelOptRule JOIN_INSTANCE =
+            Config.EMPTY.as(Config.class).withJoin().toRule();
+    public static final RelOptRule CALC_INSTANCE =
+            Config.EMPTY.as(Config.class).withCalc().toRule();
+
+    private static final UnreachableCoalesceArgumentsRemoveRexShuttle REX_SHUTTLE_INSTANCE =
+            new UnreachableCoalesceArgumentsRemoveRexShuttle();
+
+    public UnreachableCoalesceArgumentsRemoveRule(Config config) {
+        super(config);
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        final RelNode relNode = call.rel(0);
+        call.transformTo(relNode.accept(REX_SHUTTLE_INSTANCE));
+    }
+
+    private static class UnreachableCoalesceArgumentsRemoveRexShuttle extends RexShuttle {
+
+        @Override
+        public RexNode visitCall(RexCall call) {
+            call = (RexCall) super.visitCall(call);
+
+            // Not a coalesce invocation, skip it
+            if (!call.getOperator()
+                    .getName()
+                    .equals(BuiltInFunctionDefinitions.COALESCE.getName())) {
+                return call;
+            }
+
+            for (int firstNullableIndex = 0;
+                    firstNullableIndex < call.operands.size();
+                    firstNullableIndex++) {
+                RexNode argument = call.operands.get(firstNullableIndex);
+                // Check if it's non null
+                if (!argument.getType().isNullable()) {
+                    // If it's the first argument, just return the argument without the coalesce
+                    // invocation
+                    if (firstNullableIndex == 0) {
+                        return argument;
+                    }
+
+                    // If it's the last argument, no change is required
+                    if (firstNullableIndex == call.operands.size() - 1) {
+                        break;
+                    }
+
+                    // Return the coalesce invocation with a trimmed argument list
+                    List<RexNode> trimmedOperandsList =
+                            call.operands.subList(0, firstNullableIndex + 1);
+                    return call.clone(call.getType(), trimmedOperandsList);
+                }
+            }
+
+            // Either no arguments are non-null, or only the last one is non-null
+            return call;
+        }
+    }
+
+    // ---------------------------------------------------------------------------------------------
+
+    /** Configuration for {@link UnreachableCoalesceArgumentsRemoveRule}. */
+    public interface Config extends RelRule.Config {
+
+        @Override
+        default RelOptRule toRule() {
+            return new UnreachableCoalesceArgumentsRemoveRule(this);
+        }
+
+        default Config withProject() {
```

Comment:
For all the `with*` methods it'd be nice to extract the inner predicates to make the code more readable (see e.g. `PushFilterPastChangelogNormalizeRule`).